utils/src/shell_wrapper.rs

```diff
@@ -18,37 +20,76 @@ fn real_main() -> anyhow::Result<()> {
 
     // Skip if environment was already set
     if env::var_os("__NIXOS_SET_ENVIRONMENT_DONE") != Some("1".into()) {
-        // Load the environment from /etc/set-environment
-        let output = Command::new(env!("NIXOS_WSL_SH"))
-            .args(&[
-                "-c",
-                &format!(". /etc/set-environment && {} -0", env!("NIXOS_WSL_ENV")),
-            ])
-            .output()
-            .context("when reading /etc/set-environment")?;
-
-        // Parse the output
-        let output_string =
-            String::from_utf8(output.stdout).context("when decoding the output of env")?;
-        let env = output_string
-            .split('\0')
-            .filter(|entry| !entry.is_empty())
-            .map(|entry| {
-                entry
-                    .split_once("=")
-                    .ok_or(anyhow!("invalid env entry: {}", entry))
-            })
-            .collect::<Result<Vec<_>, _>>()
-            .context("when parsing the output of env")?;
-
-        // Apply the environment variables
-        for &(key, val) in &env {
-            env::set_var(key, val);
-        }
+        || -> anyhow::Result<()> {
+            if !std::path::Path::new("/etc/set-environment").exists() {
+                eprintln!("[shell-wrapper] Warning: /etc/set-environment does not exist");
+                return Ok(());
+            }
+
+            unsafe {
+                // WSL starts a single shell under login to make sure that a logind session exists.
+                // That shell is started with SIGCHLD ignored
+                // If it is, we are probably that shell and can just skip setting the environment
+                let mut act: sigaction = MaybeUninit::zeroed().assume_init();
+                sigaction(SIGCHLD, PT_NULL as *const sigaction, &mut act);
+                if act.sa_sigaction == SIG_IGN {
+                    return Ok(());
+                }
+            }
```

Comment:
https://docs.rs/nix/latest/nix/sys/signal/fn.sigaction.html

Reply:
Never mind that can't read (by passing null), only write. Should probably add a comment for that.

Reply:
added a comment